analysis.ml

```diff
@@ -228,3 +229,38 @@ let dominates (instrs : segment) : pc -> pc -> bool =
   fun pc pc' ->
     let doms = dominators pc' in
     PcSet.exists ((=) pc) doms
+
+let lifetime_analysis (instrs : segment) =
+  let introduction instr = VarSet.elements (ModedVarSet.untyped (declared_vars instr)) in
+  let elimination instr = VarSet.elements (required_vars instr) in
+  scope_analysis introduction elimination instrs
+
+(* returns a 'pc -> pc set' computing the set of instructions depending on a declaration *)
+let required (instrs : segment) : pc -> PcSet.t =
+  let res = lifetime_analysis instrs in
+  fun pc ->
+    let instr = instrs.(pc) in
+    let declared = VarSet.elements (ModedVarSet.untyped (declared_vars instr)) in
+    let required_of var = VariableMap.at var (res pc) in
+    let all_requires = List.map required_of declared in
+    List.fold_left PcSet.union PcSet.empty all_requires
```

Comment:
This works and is rather clear. If you wanted, you could easily avoid the conversion to lists and back:

```ocaml
    let add_uses (_, x) used = PcSet.union used (VariableMap.at x (res pc)) in
    ModedVarSet.fold add_uses (defined_vars instrs.(pc)) PcSet.empty
```